### PR TITLE
AB#61483 recovery of deleted workflows and dashboards

### PR DESCRIPTION
**Pull request by ketanreliefapps**
# Description
When deleting a workflow or a dashboard it would be good to be able to recover the latest version of the workflow or dashboard without it causing a lot of difficulty/disruption to the solution, in the case that it was deleted incorrectly, or, if the business has a future need to reinstate the workflow or dashboard. 
Solution:
backend: instead of deleting dashboard / workflow, archive it

## Ticket
[Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/63506)
[FrontEnd Branch](https://github.com/ReliefApplications/oort-frontend/pull/1505)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (refactor or addition to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [x] * I have performed a self-review of my own code
- [x] * I have commented my code, particularly in hard-to-understand areas
- [x] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [x] * I have included screenshots describing my changes if relevant
- [x] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/schema/mutation/deleteDashboard.mutation.ts [hunk 1]
```diff
@@ -23,22 +23,29 @@ export default {
...
-        return await Dashboard.findByIdAndDelete(args.id);
-      } else {
-        const page = await Page.accessibleBy(ability, 'delete').where({
-          content: args.id,
-        });
-        const step = await Step.accessibleBy(ability, 'delete').where({
-          content: args.id,
-        });
-        if (page || step) {
+      if (args.hardDelete) {
```
Comment (AntoineRelief) [inline src/schema/mutation/deleteDashboard.mutation.ts:27]: @ketanreliefapps not a valid arg
keep the delete, and depending on current status of the dashboard, hard delete it or not

File: src/schema/mutation/deleteWorkflow.mutation.ts [hunk 2]
```diff
@@ -25,18 +25,26 @@ export default {
...
-        workflow = await Workflow.findByIdAndDelete(args.id);
-      } else {
-        const page = await Page.accessibleBy(ability, 'delete').where({
-          content: args.id,
-        });
-        const step = await Step.accessibleBy(ability, 'delete').where({
-          content: args.id,
-        });
-        if (page || step) {
+      if (args.hardDelete) {
```
Comment (AntoineRelief) [inline src/schema/mutation/deleteWorkflow.mutation.ts:28]: @ketanreliefapps same comment there

Comment (ketanreliefapps) [reply src/schema/mutation/deleteWorkflow.mutation.ts]: Done

**Comment by MwanPygmay** [comment]
- merged beta
- changed some naming: const dashboards = dashboard.findone() does not seem logical to me.
- changed some findOne to findById when possible

File: src/security/defineUserAbility.ts [hunk 3]
```diff
@@ -341,5 +341,12 @@ export default function defineUserAbility(user: User | Client): AppAbility {
     can(['create', 'read', 'update', 'delete'], 'Layer');
   }
 
+  /* ===
+    Creation / Access / Edition / Deletion of Archived Records
+  === */
+  if (userGlobalPermissions.includes(permissions.canSeeArchived)) {
```
Comment (AntoineRelief) [inline src/security/defineUserAbility.ts:347]: @MwanPygmay 
I know you didn't write that, but when taking a ticket already worked on by someone else, please check that the changes are relevant or not
this is not

**Pull request closed without merging**